### PR TITLE
Feature validation

### PR DESCRIPTION
**Pull request by gra-moore**
Validation support
Use of EntityCollection
YAML schema parser

### CODE REVIEW

File: provider.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	"encoding/json"
+	datahub "github.com/mimiro-io/datahub-client-sdk-go"
+	egdm "github.com/mimiro-io/entity-graph-data-model"
+)
+
+type RemoteDataProvider struct {
+	client *datahub.Client
+}
+
+func NewRemoteDatahubQueryHelper(endpoint string) (*RemoteDataProvider, error) {
```
Comment (sveeig) [inline provider.go]: Constructor name and return type name inconsistent

File: validator.go [hunk 2]
```diff
@@ -1,16 +1,125 @@
...
+	violation := &ConstraintViolation{}
+	violation.Constraint = constraint
+	violation.Entity = entity
+	violation.ViolationType = violationType
+	violation.Message = message
+	return violation
+}
+
+// ValidateDataset validates the given dataset against the data in the named dataset
+func (v *Validator) ValidateDataset(schema *Schema, datasetName string) (ok bool, exceptions []*ConstraintViolation, err error) {
```
Comment (sveeig) [inline validator.go:88]: Having both bool and error returned can be confusing. Eg what does it mean if ok=true and err!=nil

File: validator.go [hunk 3]
```diff
@@ -30,61 +139,253 @@ func (v *Validator) Validate(schema *Schema, entity *egdm.Entity) (ok bool, exce
...
 	}
 
-	return false
+	return false, nil, errors.New("constraint type not supported")
 }
 
-type ValidationContext interface {
-	// return a given entity by id in the named datasets
-	GetEntity(id string, datasets []string) *egdm.Entity
+func (v *Validator) CheckReferenceConstraint(entity *egdm.Entity, constraint *ReferenceConstraint) (bool, *ConstraintViolation, error) {
```
Comment (sveeig) [inline validator.go:246]: Cleaner to just return a Validation error, which is nil when ok and contains the violation object when raised?

File: provider.go (hunk 1 above)
Comment (gra-moore) [reply provider.go]: Agreed will fix.

File: validator.go (hunk 2 above)
Comment (gra-moore) [reply validator.go:88]: decided to keep this as the errors are different from validation exceptions. the ok is a shorthand for checking the exceptions is nil. Question is valid, but not different to what does it mean if exceptions is not empty and there are errors.  As with all go, errors should take precedence if non-nil over other func return values.

File: validator.go (hunk 3 above)
Comment (gra-moore) [reply validator.go:246]: same as above.

**Review by sveeig — APPROVED** [review]

**Pull request merged**